components/script/dom/bindings/str.rs

```diff
@@ -208,6 +208,95 @@ impl DOMString {
         self.0.truncate(last_non_whitespace);
         let _ = self.0.splice(0..first_non_whitespace, "");
     }
+
+    /// Valid date string should be YYYY-MM-DD, YYYY must be four or more
+    /// https://html.spec.whatwg.org/multipage/#valid-date-string
+    pub fn is_valid_date_string(&mut self) -> bool {
+        // Step 2
+        let value = &*(self.0);
+        let value_vec: Vec<&str> = value.split('-').collect();
+        if !(value_vec.len() == 3) {
+            return false;
+        }
+        let year = value_vec[0];
+        let month = value_vec[1];
+        let day = value_vec[2];
+
+        // Step 1
+        let month_string = format!("{}-{}", year, month);
+        if !self.check_valid_month_string(&*month_string) {
+            return false;
+        }
+
+        // Step 3
+        if !(day.len() == 2 && day.chars().all(|c| c.is_digit(10))) {
+            return false;
+        }
+        let year_int = year.parse::<i32>().unwrap();
+        let month_int = month.parse::<i32>().unwrap();
+        let day_int = day.parse::<i32>().unwrap();
+        if day_int == 0 {
+            return false;
+        } else if day_int > self.max_day_in_month(year_int, month_int) {
+            return false;
+        }
+        return true;
+    }
+
+    /// Valid month string should be "YYYY-MM"
+    /// https://html.spec.whatwg.org/multipage/#valid-month-string
+    pub fn is_valid_month_string(&mut self) -> bool {
```

Comment:
What's the purpose for this function? It looks to me as an unnecessary indirection.

Reply:
@KiChjang 
because the origin usage is:
```
DOMString.is_valid_month_string()
```

The logic in it can be use in `is_valid_date_string()`

Reply:
So perhaps my question is better phrased as "why is `is_valid_month_string` and `check_valid_month_string`" not the same function?

Reply:
@KiChjang  
when we call `is_valid_month_string`  will deal the `self.0` string.

But when we call  `is_valid_date_string` and use logic in `is_valid_month_string`, we cannot directly call `is_valid_month_string` in `is_valid_date_string`, because at that time `self.0` might be `1990-12-90` and `is_valid_month_string` will return `false`.

That's why I make the logic out to `check_valid_month_string`.

Maybe there is a better way to do this.

Reply:
I think we need to be careful when reading the spec here - according to my reading, the spec does not fail on [parsing a month component](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#parse-a-month-component) if the string contains a date at the end; it simply does not parse the rest, and returns _year_ and _month_ to the caller of this algorithm.

Reply:
@KiChjang 
According to the spec:

A string is a valid month string representing a year year and month month if it consists of the following components in the given order:

1. Four or more ASCII digits, representing year, where year > 0
2.  A U+002D HYPHEN-MINUS character (-)
3. Two ASCII digits, representing the month month, in the range 1 ≤ month ≤ 12

Which means only "YYYY-MM" is valid

Reply:
Note that it says "consists of the following components" - it's missing the *only* keyword. In other words, date strings are also valid month strings.

Furthermore, I believe we have agreed to use the `parse a month string` algorithm instead of the validation algorithm, so this part of the spec shouldn't be relevant to your implementation.